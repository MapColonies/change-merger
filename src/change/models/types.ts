import { OsmChange } from '@map-colonies/node-osm-elements';

interface IdMapping {
  externalId: string;
  tempOsmId: number;
}

interface OsmXmlTag {
  k: string;
  v: string;
}

interface ChangeWithMetadata {
  externalId: string;
  change: OsmChange;
  tempOsmId?: number;
  action: 'create' | 'modify' | 'delete';
}

interface OsmXmlBase {
  id: number;
  version: number;
  changeset: number;
  tag: OsmXmlTag[];
}

interface OsmXmlNode extends OsmXmlBase {
  lon: number;
  lat: number;
}

interface OsmXmlWay extends OsmXmlBase {
  nd: { ref: number }[];
}

interface ChangeActionObj {
  node: OsmXmlNode[];
  way: OsmXmlWay[];
}

 interface OsmXmlChange {
  version: '0.6';
  generator: string;
  create: ChangeActionObj;
  modify: ChangeActionObj;
  delete: ChangeActionObj;
}


export { OsmXmlNode, OsmXmlWay, OsmXmlTag, OsmXmlChange, ChangeWithMetadata, IdMapping };
