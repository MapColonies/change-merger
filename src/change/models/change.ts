import { OsmChange } from '@map-colonies/node-osm-elements';

interface OsmXmlBase {
  id: number;
  version: number;
  changeset: number;
  tag?: OsmXmlTag | OsmXmlTag[];
}

interface ChangeActionObj {
  node: OsmXmlNode[];
  way: OsmXmlWay[];
}

interface NodeChange {
  node: OsmXmlNode | OsmXmlNode[];
}

interface WayChange {
  way: OsmXmlWay | OsmXmlWay[];
}

interface RelationChange {
  relation: unknown | unknown[];
}

export interface OsmXmlTag {
  k: string;
  v: string;
}

export interface OsmXmlNode extends OsmXmlBase {
  lon: number;
  lat: number;
}

export interface OsmXmlWay extends OsmXmlBase {
  nd: { ref: number }[];
}

export type ElementChange = NodeChange | WayChange | RelationChange;

export interface ChangeWithMetadata {
  externalId: string;
  change: OsmChange;
  tempOsmId?: number;
  action: 'create' | 'modify' | 'delete';
}

export interface MergedOsmXmlChange {
  version: '0.6';
  generator: string;
  create: ChangeActionObj;
  modify: ChangeActionObj;
  delete: ChangeActionObj;
}

export interface OsmXmlChange {
  version: '0.6';
  generator: string;
  create?: ElementChange | ElementChange[];
  modify?: ElementChange | ElementChange[];
  delete?: ElementChange | ElementChange[];
}
