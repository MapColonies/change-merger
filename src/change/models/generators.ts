import { OsmNode, OsmWay } from '@map-colonies/node-osm-elements';
import { OsmXmlNode, OsmXmlTag, OsmXmlWay } from './change';

const createOsmXmlTag = (oldTags?: Record<string, string>): OsmXmlTag[] => {
  if (!oldTags) {
    return [];
  }
  return Object.entries(oldTags).map(([k, v]) => ({ k, v }));
};

const createOsmXmlNode = (osmNode: OsmNode, changesetId: number): OsmXmlNode => {
  const { type, tags, ...node } = osmNode as Required<OsmNode>;
  return { ...node, changeset: changesetId, tag: createOsmXmlTag(tags) };
};

const createOsmXmlWay = (osmWay: OsmWay, tempMapping: Map<number, number>, changesetId: number): OsmXmlWay => {
  const { tags, nodes, type, ...rest } = osmWay as Required<OsmWay>;
  return {
    ...rest,
    changeset: changesetId,
    nd: nodes.map((node) => ({ ref: tempMapping.get(node.id) ?? node.id })),
    tag: createOsmXmlTag(tags),
  };
};

export { createOsmXmlNode, createOsmXmlWay, createOsmXmlTag };
