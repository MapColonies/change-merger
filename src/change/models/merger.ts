import { BaseElement, OsmNode, OsmWay } from '@map-colonies/node-osm-elements';
import { SERVICE_NAME } from '../../common/constants';
import { IdGenerator } from '../utils/idGenerator';
import { createOsmXmlNode, createOsmXmlWay } from './generators';
import { IdMapping, ChangeWithMetadata, OsmXmlChange, OsmXmlNode, OsmXmlWay } from './types';

const isNode = (element: BaseElement): element is OsmNode => element.type === 'node';
const isWay = (element: BaseElement): element is OsmWay => element.type === 'way';

const handleChangeArray = (
  elements: BaseElement[],
  idGenerator: IdGenerator,
  tempMapping: Map<number, number>,
  changesetId: number
): [OsmXmlNode[], OsmXmlWay?] => {
  const nodes: OsmXmlNode[] = [];
  let osmWay: OsmWay | null = null;

  for (let i = 0; i < elements.length; i++) {
    const element = elements[i];
    if (element.id < 0) {
      const newId = idGenerator.getId();
      tempMapping.set(element.id, newId);
      element.id = newId;
    }

    if (isNode(element)) {
      nodes.push(createOsmXmlNode(element, changesetId));
    } else if (isWay(element)) {
      osmWay = element;
    }
  }

  let xmlWay: OsmXmlWay | undefined = undefined;
  if (osmWay) {
    xmlWay = createOsmXmlWay(osmWay, tempMapping, changesetId);
  }
  return [nodes, xmlWay];
};

const handleChangeObj = (changeObj: ChangeWithMetadata, idGenerator: IdGenerator, change: OsmXmlChange, changesetId: number): number => {
  const tempMapping = new Map<number, number>();

  const actions: typeof changeObj.action[] = ['create', 'modify', 'delete'];

  actions.forEach((action) => {
    const elementArr = changeObj.change[action];
    if (elementArr) {
      const [nodes, way] = handleChangeArray(elementArr, idGenerator, tempMapping, changesetId);
      change[action].node.push(...nodes);
      if (way) {
        change[action].way.push(way);
      }
    }
  });

  return changeObj.tempOsmId !== undefined ? (tempMapping.get(changeObj.tempOsmId) as number) : 0;
};

const mergeChanges = (changes: ChangeWithMetadata[], changesetId: number): [OsmXmlChange, IdMapping[], string[]] => {
  const idGenerator = new IdGenerator();
  const change: OsmXmlChange = {
    generator: SERVICE_NAME,
    version: '0.6',
    create: { node: [], way: [] },
    modify: { node: [], way: [] },
    delete: { way: [], node: [] },
  };
  const idsToAdd: IdMapping[] = [];
  const idsToDelete: string[] = [];

  for (let i = 0; i < changes.length; i++) {
    const changeObj = changes[i];
    const tempOsmId = handleChangeObj(changeObj, idGenerator, change, changesetId);

    if (changeObj.action === 'create') {
      idsToAdd.push({ tempOsmId, externalId: changeObj.externalId });
    }

    if (changeObj.action === 'delete') {
      idsToDelete.push(changeObj.externalId);
    }
  }

  return [change, idsToAdd, idsToDelete];
};

export { mergeChanges };
