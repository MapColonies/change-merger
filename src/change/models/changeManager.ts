import { Logger } from '@map-colonies/js-logger';
import { inject, injectable } from 'tsyringe';
import { ConfigType } from '@src/common/config';
import { OsmElementType } from '@map-colonies/node-osm-elements';
import { SERVICES } from '../../common/constants';
import { convertToXml } from '../utils/xml';
import { mergeChanges } from './merger';
import { ChangeWithMetadata, ElementChange, OsmXmlChange, OsmXmlNode, OsmXmlWay } from './change';
import { IdMapping, InterpretAction, InterpretedMapping, InterpretResult } from './types';

@injectable()
export class ChangeManager {
  public constructor(
    @inject(SERVICES.LOGGER) private readonly logger: Logger,
    @inject(SERVICES.CONFIG) private readonly config: ConfigType
  ) {}

  public mergeChanges(changes: ChangeWithMetadata[], changesetId: number): [string, IdMapping[], string[]] {
    this.logger.info({ msg: 'started changes merging', count: changes.length });

    const [change, idsToCreate, idsToDelete] = mergeChanges(changes, changesetId);
    return [convertToXml({ osmChange: change }), idsToCreate, idsToDelete];
  }

  public interpretChange(change: OsmXmlChange, actions?: InterpretAction[]): Partial<InterpretResult> {
    let created, deleted;

    this.logger.info({ msg: 'started change interpret', actions });

    // due to xml parsing convertion possibly converting a single item array to just the item, we'll safely check beforehand
    if (!actions || actions.includes('create')) {
      created = change.create ? (Array.isArray(change.create) ? this.interpret(change.create) : this.interpret([change.create])) : [];
    }

    if (!actions || actions.includes('delete')) {
      deleted = change.delete ? (Array.isArray(change.delete) ? this.interpret(change.delete) : this.interpret([change.delete])) : [];
    }

    return { created, deleted };
  }

  private interpret(elements: ElementChange[]): InterpretedMapping[] {
    const mapping: InterpretedMapping[] = [];

    elements.forEach((wrappedElement) => {
      // skip relations
      if ('relation' in wrappedElement) {
        return;
      }

      // determine if node or way
      let element: OsmXmlNode | OsmXmlWay;
      let type: OsmElementType;
      if ('node' in wrappedElement) {
        element = wrappedElement.node;
        type = 'node';
      } else {
        element = wrappedElement.way;
        type = 'way';
      }

      // skip element with no tags
      if (element.tag === undefined) {
        return;
      }

      // if found, add to mapping the osmId, externalId pair
      const tags = Array.isArray(element.tag) ? element.tag : [element.tag];
      const externalIdTag = tags.find((tag) => tag.k === this.config.get('app.externalIdTag'));
      if (externalIdTag) {
        mapping.push({ type, osmId: +element.id, externalId: externalIdTag.v });
      }
    });

    return mapping;
  }
}
