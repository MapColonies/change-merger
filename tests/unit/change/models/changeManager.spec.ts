import { ChangeManager } from '../../../../src/change/models/changeManager';
import { getSampleData } from '../../../sampleData';

describe('changeManager', function () {
  describe('#mergeChanges', function () {
    it('should merge the changes into the same result', function () {
      const manager = new ChangeManager({ log: jest.fn() });
      const changes = getSampleData();

      const change = manager.mergeChanges(changes, 2);

      expect(change).toMatchSnapshot();
    });

    it('should encode strings using html encoding', function () {
      const manager = new ChangeManager({ log: jest.fn() });
      const changes = getSampleData();
      const tags = changes[0].change.create?.[0].tags;
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      tags!.html = '< > " \' & Â© âˆ† ðŸ„¯';

      const change = manager.mergeChanges(changes, 2);

      expect(change).toMatchSnapshot();
    });
  });
});
