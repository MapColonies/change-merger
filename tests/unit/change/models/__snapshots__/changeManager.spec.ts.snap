// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changeManager #mergeChanges should merge the changes into the same result 1`] = `
Array [
  "<osmChange generator=\\"change-merger\\" version=\\"0.6\\"><create><node id=\\"-1\\" version=\\"0\\" lon=\\"24\\" lat=\\"24\\" changeset=\\"2\\"><tag k=\\"cat\\" v=\\"bark\\"></tag></node><node id=\\"-2\\" version=\\"0\\" lon=\\"25\\" lat=\\"25\\" changeset=\\"2\\"></node><node id=\\"-3\\" version=\\"0\\" lon=\\"26\\" lat=\\"26\\" changeset=\\"2\\"></node><node id=\\"-5\\" version=\\"0\\" lon=\\"24\\" lat=\\"24\\" changeset=\\"2\\"><tag k=\\"cat\\" v=\\"bark\\"></tag></node><way id=\\"-4\\" version=\\"0\\" changeset=\\"2\\"><nd ref=\\"-1\\"></nd><nd ref=\\"-2\\"></nd><nd ref=\\"-3\\"></nd></way></create><modify><way id=\\"1\\" version=\\"0\\" changeset=\\"2\\"><nd ref=\\"-5\\"></nd><nd ref=\\"3\\"></nd><nd ref=\\"4\\"></nd></way></modify><delete><way id=\\"5\\" version=\\"0\\" changeset=\\"2\\"><nd ref=\\"6\\"></nd><nd ref=\\"7\\"></nd><nd ref=\\"8\\"></nd></way><node id=\\"1\\" version=\\"0\\" lon=\\"24\\" lat=\\"24\\" changeset=\\"2\\"><tag k=\\"cat\\" v=\\"bark\\"></tag></node><node id=\\"6\\" version=\\"0\\" lon=\\"24\\" lat=\\"24\\" changeset=\\"2\\"><tag k=\\"cat\\" v=\\"bark\\"></tag></node><node id=\\"7\\" version=\\"0\\" lon=\\"25\\" lat=\\"25\\" changeset=\\"2\\"></node><node id=\\"8\\" version=\\"0\\" lon=\\"26\\" lat=\\"26\\" changeset=\\"2\\"></node></delete></osmChange>",
  Array [
    Object {
      "externalId": "a",
      "tempOsmId": -4,
    },
  ],
  Array [
    "c",
  ],
]
`;
